prow/tide/tide.go

```diff
@@ -231,17 +243,38 @@ func (sc *statusController) setStatuses(all, pool []PullRequest) {
 
 func (sc *statusController) run() {
 	lastStart := time.Time{} // Zero value so that we don't wait the first loop.
-	// wait for a new pool
-	for range sc.newPoolPending {
+	poolPending := false
+RunLoop:
+	for {
+		// wait for a new pool
+		if poolPending {
+			poolPending = false
+		} else if !<-sc.newPoolPending {
+			// chan was closed
+			break
+		}
 		// wait for the min sync period time to elapse if needed.
-		time.Sleep(time.Until(lastStart.Add(sc.ca.Config().Tide.StatusUpdatePeriod)))
-		lastStart = time.Now()
-
-		sc.Lock()
-		pool := sc.poolPRs
-		sc.Unlock()
-		sc.sync(pool)
+		wait := time.After(time.Until(lastStart.Add(sc.ca.Config().Tide.StatusUpdatePeriod)))
+	WaitLoop:
+		for {
+			select {
+			case <-wait:
+				lastStart = time.Now()
+				sc.Lock()
+				pool := sc.poolPRs
+				sc.Unlock()
+				sc.sync(pool)
+				break WaitLoop
```

Comment:
The `break` will exit out of the `select` block instead of the `for` loop without this: https://golang.org/ref/spec#Break_statements
This won't be needed once I change this to a separate function as you recommended though.